src/stdune/temp.ml

```diff
@@ -1,16 +1,69 @@
+let try_times n ~f =
+  assert (n > 0);
+  let rec loop n =
+    if n = 1 then
+      f n
+    else
+      match f n with
+      | exception _ -> loop (n - 1)
+      | r -> r
+  in
+  loop n
+
+let prng = lazy (Random.State.make_self_init ())
+
+let temp_file_name ?temp_dir ~prefix ~suffix () =
+  let temp_dir =
+    match temp_dir with
+    | None -> Filename.get_temp_dir_name ()
+    | Some t -> t
+  in
+  let rnd = Random.State.bits (Lazy.force prng) land 0xFFFFFF in
+  Filename.concat temp_dir (Printf.sprintf "%s%06x%s" prefix rnd suffix)
+
 let tmp_files = ref Path.Set.empty
 
+let tmp_dirs = ref Path.Set.empty
+
+external open_desc : string -> open_flag list -> int -> int = "caml_sys_open"
+
+external close_desc : int -> unit = "caml_sys_close"
+
+let create_temp_file name =
+  close_desc (open_desc name [ Open_wronly; Open_creat; Open_excl ] 0o600)
+
+let create_temp_dir name =
+  match Path.Fpath.mkdir_p name with
+  | Created -> ()
+  | Already_exists -> failwith "temp dir exists"
```

Comment:
We should raise `Unix_error (ENOENT, "mkdir", name)` here, that would produce a better error message and would be more in line with the file case.